### PR TITLE
feat: server side rendering

### PR DESCRIPTION
**Pull request by mthuret**
This PR add the support of server side rendering through a new API:

`findResults` to be able to get the results from an InstantSearch instance. 
a new kind of `InstantSearch` instance (/server), that can take a "resultsState" props. 

TODO: 

- [x] Improve the SSR guide.
- [x] Check if static getSearchParameters can be removed.
- [x] add the next-app on our example page. 
- [x] `findResults` name could be better. any ideas? (@Haroenv propose `findResultsState`).

### CODE REVIEW

File: docgen/src/guide/Server-side_rendering.md [hunk 1]
```diff
@@ -6,8 +6,57 @@ category: guide
 navWeight: 30
 ---
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widgets declared. That's why at the first rendering there is no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits.
```
Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: * every widget declared

Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: * there are no results yet

File: docgen/src/guide/Server-side_rendering.md [hunk 2]
```diff
@@ -6,8 +6,57 @@ category: guide
 navWeight: 30
 ---
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widgets declared. That's why at the first rendering there is no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits. 
+
+If you want to have a server side rendering application using React InstantSearch you will need to use the special API entries we provide for this particular use-case:
```
Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: you can use

File: docgen/src/guide/Server-side_rendering.md [hunk 3]
```diff
@@ -6,8 +6,57 @@ category: guide
...
 ---
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widgets declared. That's why at the first rendering there is no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits. 
+
+If you want to have a server side rendering application using React InstantSearch you will need to use the special API entries we provide for this particular use-case: 
+
+* a new type of `<InstantSearch>` available under `'react-instantsearch/server'`
+* a `findResults` function available under `'react-instantsearch/server'`
```
Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: * is available

File: docgen/src/guide/Server-side_rendering.md [hunk 4]
```diff
@@ -6,8 +6,57 @@ category: guide
...
+        onSearchStateChange={this.props.onSearchStateChange}
+        searchState={this.props.searchState}
+      >
+        <Hits />
+      </InstantSearch>
+    );
+  }
+}
+
+const resultsState = await findResults(App, { searchState });
```
Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: Where is `searchState` defined?

File: docgen/src/guide/Server-side_rendering.md [hunk 5]
```diff
@@ -6,8 +6,57 @@ category: guide
...
+```
+
+Then, once you'll get those results, you can actually give it as a prop to your `<InstantSearch/>` instance. You'll need a mechanism to render your app after fetching those results.
```
Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: I would add a few more lines here that actually show how it's done in the example.

File: packages/react-instantsearch/scripts/build.sh [hunk 6]
```diff
@@ -23,6 +23,7 @@ babel -q index.js -o dist/index.js &&
 babel -q dom.js -o dist/dom.js &&
 babel -q connectors.js -o dist/connectors.js &&
 babel -q native.js -o dist/native.js
```
Comment (Haroenv) [inline packages/react-instantsearch/scripts/build.sh]: you forgot an `&&` here

File: packages/react-instantsearch/src/core/InstantSearch.js [hunk 7]
```diff
@@ -30,6 +30,7 @@ function validateNextProps(props, nextProps) {
  * @propType {func} [onSearchStateChange] - Function to be called everytime a new search is done. Useful for [URL Routing](guide/Routing.html).
  * @propType {object} [searchState] - Object to inject some search state. Switches the InstantSearch component in controlled mode. Useful for [URL Routing](guide/Routing.html).
  * @propType {func} [createURL] - Function to call when creating links, useful for [URL Routing](guide/Routing.html).
+ * @propType {object} [resultsState] - Object or Array to inject some results to be used at first rendering. Useful for [Server Side Rendering](guide/Server-side_rendering.html). This object or array comes from the `findResults` function.
```
Comment (Haroenv) [inline packages/react-instantsearch/src/core/InstantSearch.js]: is it an array when multi index, and object when single index? Might want to elaborate a little on that here.

File: packages/react-instantsearch/src/core/createConnector.js [hunk 8]
```diff
@@ -52,6 +51,7 @@ export default function createConnector(connectorDesc) {
       static defaultClassNames = Composed.defaultClassNames;
       static propTypes = connectorDesc.propTypes;
       static defaultProps = connectorDesc.defaultProps;
+      static getSearchParameters = connectorDesc.getSearchParameters; //can be removed?
```
Comment (Haroenv) [inline packages/react-instantsearch/src/core/createConnector.js]: Why is it added if it can be removed?

File: packages/react-instantsearch/src/core/createConnector.js [hunk 9]
```diff
@@ -65,12 +65,15 @@ export default function createConnector(connectorDesc) {
         const { ais: { store, widgetsManager }, multiIndexContext } = context;
         this.state = {
           props: this.getProvidedProps(props),
+          canRender: false, //indicate if component are rendered on a server or on an client.
```
Comment (Haroenv) [inline packages/react-instantsearch/src/core/createConnector.js]: indicates if a component is rendered with `findResults` (true) or not

File: packages/react-instantsearch/src/core/createConnector.test.js [hunk 10]
```diff
@@ -381,6 +383,54 @@ describe('createConnector', () => {
...
+        displayName: 'CoolConnector',
+        getProvidedProps: () => null,
+        getId,
+      })(() => null);
+
+      mount(<Connected {...props} />, {
+        context,
+      });
+
+      expect(onSearchParameters.mock.calls.length).toBe(1);
```
Comment (Haroenv) [inline packages/react-instantsearch/src/core/createConnector.test.js]: You can use [`.toHaveBeenCalledTimes(1)`](http://facebook.github.io/jest/docs/expect.html#tohavebeencalledtimesnumber)

File: packages/react-instantsearch/src/core/createInstantSearchServer.js [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        { context: searchParameter.context },
+        acc[index] ? acc[index] : sharedSearchParameters,
+        searchParameter.props,
+        searchParameter.searchState
+      );
+      acc[index] = sp;
+      return acc;
+    }, {});
+
+  let search;
```
Comment (Haroenv) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: can this be refactored to not be reassigned?

File: packages/react-instantsearch/src/core/createInstantSearchServer.test.js [hunk 12]
```diff
@@ -0,0 +1,186 @@
+/* eslint-env jest, jasmine */
+/* eslint-disable no-console */
```
Comment (Haroenv) [inline packages/react-instantsearch/src/core/createInstantSearchServer.test.js]: can be removed

File: packages/react-instantsearch/src/core/createConnector.js (hunk 8 above)
Comment (mthuret) [inline packages/react-instantsearch/src/core/createConnector.js]: Indeed I'm wondering if I should mention this. It's not meant to be used without the usage of "findResults".

File: packages/react-instantsearch/src/core/InstantSearch.js (hunk 7 above)
Comment (mthuret) [inline packages/react-instantsearch/src/core/InstantSearch.js]: Indeed I'm wondering if I should mention this. It's not meant to be used without the usage of "findResults".

File: packages/react-instantsearch/src/core/createConnector.test.js (hunk 10 above)
Comment (mthuret) [inline packages/react-instantsearch/src/core/createConnector.test.js]: Will do this refactor in another PR (impact a lot of files)

Comment (Haroenv) [inline packages/react-instantsearch/src/core/createConnector.test.js]: sure, just saw it in this diff, still useful to know 😄

Comment (mthuret) [inline packages/react-instantsearch/src/core/createConnector.test.js]: Yep sure, way more readable :)

Comment (mthuret) [inline packages/react-instantsearch/src/core/createConnector.test.js]: At some point, I'll switch to toHaveBeenCalledWith also.

File: docgen/src/guide/Server-side_rendering.md (hunk 5 above)
Comment (mthuret) [inline docgen/src/guide/Server-side_rendering.md]: I think in the end, I should remove the code sample. Right now, it's partial so not very useful, and if I go for the full one then:

- it's next.js only
- it's complex
- it will be static and as soon as the next-app demo will update, I'm sure it will be forgotten. 

Maybe that the link to the next.js example is enough. What do you think?

Comment (Haroenv) [inline docgen/src/guide/Server-side_rendering.md]: I like this example, maybe all it lacks is a second step where you use the fetched parameters to renderToString. That would be minimal, but still viable

Comment (mthuret) [inline docgen/src/guide/Server-side_rendering.md]: I'm not sure to follow here :) The part that is missing is the render of InstantSearch using `resultsState` as prop. 

The need of `renderToString` is hidden under findResults

**Comment by harshmaur** [comment]
@marielaures You can use something like 

`getServerResults` or `findServerResults` or `getResultsFromState` instead of `findResults`

**Comment by harshmaur** [comment]
Also let me know, I can improve the docs for you. I want to get this merged quickly. 😅

File: .babelrc [hunk 13]
```diff
@@ -4,19 +4,33 @@
...
       "plugins": [
+        ["transform-runtime", {
+          "helpers": false,
+          "polyfill": false,
+          "regenerator": true
+        }],
         "transform-react-jsx-source",
-        "react-hot-loader/babel"
+        "react-hot-loader/babel",
+        "transform-regenerator",
```
Comment (vvo) [inline .babelrc]: What's the requirement for this? (Curious)

File: docgen/src/guide/Server-side_rendering.md [hunk 14]
```diff
@@ -6,8 +6,24 @@ category: guide
 navWeight: 30
 ---
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch builds the Algolia query based on every widget declared. That's why at the first rendering there are no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits.
```
Comment (vvo) [inline docgen/src/guide/Server-side_rendering.md]: I feel like we should start by saying "React InstantSearch is compatible with server-side rendering, here are examples on how to use it:".

Then at the end of the page add a paragraph explaining a bit the inner details like this first sentence.

Otherwise I open this page, I see a sentence that seem negative, I think "well maybe doesn't work, let's move on".

File: docgen/src/guide/Server-side_rendering.md [hunk 15]
```diff
@@ -6,8 +6,24 @@ category: guide
...
+React InstantSearch builds the Algolia query based on every widget declared. That's why at the first rendering there are no results yet. This is inconvenient when doing server side rendering as the html generated by the server will not contain any hits. 
+
+If you want to have a server side rendering application using React InstantSearch you can use the special API entries we provide for this particular use-case: 
+
+* a new type of `<InstantSearch>` is available under `'react-instantsearch/server'`
+* a `findResults` function is available under `'react-instantsearch/server'`
+
+Let's see how it works!
+
+First you will need to get the first results using the `findResults` function. This function takes two parameters:
```
Comment (vvo) [inline docgen/src/guide/Server-side_rendering.md]: Either show a very small example or directly link to the example, otherwise this level of detail at this place is too much and I am not yet ready to read this/understand this.

File: docgen/src/guide/Server-side_rendering.md (hunk 14 above)
Comment (vvo) [inline docgen/src/guide/Server-side_rendering.md]: I would host the app on zeit (or any fancy host), then link it from here "Live demo", "Code".

After this, some paragraph about how it's working and how to implement it without zeit (small example, high level example).

And then some details about why it's done like that (first render undefined, declarative API handles state etc..)

File: packages/react-instantsearch/examples/next-app/.gitignore [hunk 16]
```diff
@@ -0,0 +1,19 @@
+# See https://help.github.com/ignore-files/ for more about ignoring files.
+
+# dependencies
+/node_modules
```
Comment (vvo) [inline packages/react-instantsearch/examples/next-app/.gitignore]: Why start the gitignore with a slash? (curious)

File: packages/react-instantsearch/src/core/createInstantSearch.js [hunk 17]
```diff
@@ -25,6 +25,8 @@ export default function createInstantSearch(defaultAlgoliaClient, root) {
       createURL: PropTypes.func,
       searchState: PropTypes.object,
       onSearchStateChange: PropTypes.func,
+      onSearchParameters: PropTypes.func,
+      resultsState: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
```
Comment (vvo) [inline packages/react-instantsearch/src/core/createInstantSearch.js]: Why can it be an object or array?

File: packages/react-instantsearch/src/core/createInstantSearchServer.js [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+import algoliasearchHelper, {
+  SearchResults,
+  SearchParameters,
+} from 'algoliasearch-helper';
+import ReactDom from 'react-dom/server';
+import { getIndex, hasMultipleIndex } from './indexUtils';
+import { isEmpty } from 'lodash';
+import cis from './createInstantSearch';
+
+const searchParameters = [];
```
Comment (vvo) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: Isn't there any issue having those variables global to the module?

File: packages/react-instantsearch/src/core/createInstantSearchServer.js [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+ * Creates a specialized root InstantSearch component. It accepts
+ * an algolia client and a specification of the root Element. This is the version dedicated to
+ * server side rendering. 
+ * @param {function} defaultAlgoliaClient - a function that builds an Algolia client
+ * @param {object} root - the defininition of the root of an InstantSearch sub tree.
+ * @returns {object} an InstantSearch root
+ */
+
+const createInstantSearch = function(defaultAlgoliaClient, root) {
+  const InstantSearch = cis(defaultAlgoliaClient, root);
```
Comment (vvo) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: Let's try to scope client and indexName to this createInstantSearch method. And then forward client and indexName to the other methods when needed.

File: packages/react-instantsearch/src/core/createInstantSearchServer.js [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+    getSearchParameters,
+    context,
+    props,
+    searchState,
+    index,
+  });
+};
+
+const findResults = function(App, params) {
+  ReactDom.renderToString(<App {...params} />);
```
Comment (vvo) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: Let's try to access client and indexName here:

const element = <App {...params} />;
console.log(element.client/indexName)

File: packages/react-instantsearch/src/core/createInstantSearchManager.js [hunk 21]
```diff
@@ -39,7 +40,7 @@ export default function createInstantSearchManager({
   const store = createStore({
     widgets: initialState,
     metadata: [],
-    results: null,
+    results: resultsState ? resultsState : null,
```
Comment (vvo) [inline packages/react-instantsearch/src/core/createInstantSearchManager.js]: resultsState || null

File: .babelrc (hunk 13 above)
Comment (mthuret) [inline .babelrc]: The usage of await/async in the next app example :)

File: packages/react-instantsearch/examples/next-app/.gitignore (hunk 16 above)
Comment (mthuret) [inline packages/react-instantsearch/examples/next-app/.gitignore]: Actually I don't know, it was created by the create-next-app.

File: packages/react-instantsearch/src/core/createInstantSearch.js (hunk 17 above)
Comment (mthuret) [inline packages/react-instantsearch/src/core/createInstantSearch.js]: I'm not sure to get it, but it's basically depending if you're using it with mono or multi index.

File: packages/react-instantsearch/src/core/createInstantSearchServer.js (hunk 18 above)
Comment (mthuret) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: As I'm aware of, everything is sequential so shouldn't. I tried it with several use of <InstantSearch/> (use case with several instance in the same page).

File: packages/react-instantsearch/src/core/createInstantSearchServer.js (hunk 20 above)
Comment (mthuret) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: I'll try to see if it's possible yep.

File: packages/react-instantsearch/src/core/createInstantSearchServer.js (hunk 18 above)
Comment (vvo) [inline packages/react-instantsearch/src/core/createInstantSearchServer.js]: Do you mean everything is synchronous server side wise in all situations of React server side rendering? We just need to ensure searchParameters, indexName and client are write/read in a synchronous mode and that there's no way for them to be erased/replaced at a moment we do not expect like:
- app1 server side render write variable
- app2 server sider render write variables
- app1 server side render tries to read variables
- app1 gets app2 variables
- 🏡 on 🔥

File: docgen/src/guide/Server-side_rendering.md (hunk 14 above)
Comment (mthuret) [inline docgen/src/guide/Server-side_rendering.md]: Will add the live-demo after the release of SSR.

**Comment by mthuret** [comment]
I believe we are almost done => @Haroenv and @vvo if you want to review this one more time :)

Changes since last time: 

- no more global variables under createInstantSearchServer
- better guide
- new server side rendering example (without next.js)
- API entry for `findResultsState`.

File: packages/react-instantsearch/src/core/findResultsState.js [hunk 22]
```diff
@@ -0,0 +1,16 @@
+/**
```
Comment (mthuret) [inline packages/react-instantsearch/src/core/findResultsState.js]: This is the only part I found really annoying. Couldn't succeed to add some js-doc directly under `createInstantSearchServer`. Ended up writing the findResults doc here.

Comment (vvo) [inline packages/react-instantsearch/src/core/findResultsState.js]: Legacy js doc module issues, we will have to replace it as soon as it will get too much in our way (its parser is not really comfortable with recent ways of coding).

File: docgen/src/guide/Server-side_rendering.md [hunk 23]
```diff
@@ -6,8 +6,67 @@ category: guide
 navWeight: 30
 ---
 
-React InstantSearch is not yet compatible with server-side rendering. This is one of our priority topics
-right now. We will soon provide API entries to handle this usecase.
+React InstantSearch is compatible with server-side rendering. 
+
+We provide two API entries to handle it: 
+
+* a new type of [`<InstantSearch>`](widgets/InstantSearch.html) available under `'react-instantsearch/server'`
```
Comment (vvo) [inline docgen/src/guide/Server-side_rendering.md]: This is no more the right explanation I believe?

File: docgen/src/guide/Server-side_rendering.md [hunk 24]
```diff
@@ -6,8 +6,67 @@ category: guide
...
+server.get('/', async (req, res) => {
+  const resultsState = await findResultsState(App, {searchState});
+
+  //[...]
+});
+```
+
+3. Forward this `resultState` as a prop of your `<InstantSearch>` instance. 
+
+```jsx
```
Comment (vvo) [inline docgen/src/guide/Server-side_rendering.md]: App change should come first before findResult usage. It must also include the new import statement

File: docgen/src/guide/Server-side_rendering.md [hunk 25]
```diff
@@ -6,8 +6,67 @@ category: guide
...
+* a `findResultsState` function available under `'react-instantsearch/server'`
+
+Here are the steps you need to follow to migrate:
+
+1. Import [`<InstantSearch>`](widgets/InstantSearch.html) from `'react-instantsearch/server'` instead of `'react-instantsearch/dom'`. 
+
+2. Use [`findResultsState`](server-side-renderings/findResultsState.html) to get a `resultsState` server side. 
+
+```jsx
+server.get('/', async (req, res) => {
```
Comment (vvo) [inline docgen/src/guide/Server-side_rendering.md]: Must include the import statement to get the App

**Comment by harshmaur** [comment]
Yay!!!

**Comment by vvo** [comment]
@harshmaur Will be released on Monday July 10th
